lib/sanbase_web/graphql/context_plug.ex

```
@@ -3,33 +3,76 @@ defmodule SanbaseWeb.Graphql.ContextPlug do
+
   def init(opts), do: opts
 
   def call(conn, _) do
-    context = build_context(conn)
+    context = build_context(conn, @auth_methods)
     put_private(conn, :absinthe, %{context: context})
   end
 
-  defp build_context(conn) do
+  defp build_context(conn, [auth_method|rest]) do
+    auth_method.(conn)
+    |> case do
+      :skip -> build_context(conn, rest)
+      auth -> %{auth: auth}
+    end
+  end
+
+  defp build_context(conn, []), do: %{}
+
+  def bearer_authentication(conn) do
     with ["Bearer " <> token] <- get_req_header(conn, "authorization"),
-    {:ok, current_user} <- authorize(token) do
-      %{current_user: current_user}
+    {:ok, current_user} <- bearer_authorize(token) do
+      %{auth_method: :user_token, current_user: current_user}
     else
-      _ -> %{}
+      _ -> :skip
     end
   end
 
-  defp authorize(token) do
+  def basic_authentication(conn) do
+    with ["Basic " <> auth_attempt] <- get_req_header(conn, "authorization"),
+    {:ok, current_user} <- basic_authorize(auth_attempt) do
+      %{auth_method: :basic, current_user: current_user}
+    else
+      _ -> :skip
+    end
+  end
+
+  defp bearer_authorize(token) do
     with {:ok, %User{salt: salt} = user, %{"salt" => salt}} <- SanbaseWeb.Guardian.resource_from_token(token) do
       {:ok, user}
     else
       _ ->
-        Logger.warn("Invalid token in request: #{token}")
+        Logger.warn("Invalid bearer token in request: #{token}")
         {:error, :invalid_token}
     end
   end
+
+  defp basic_authorize(auth_attempt) do
+    username = config(:basic_auth_username)
+    password = config(:basic_auth_password)
+
+    Base.encode64(username <> ":" <> password)
+    |> case do
+      ^auth_attempt -> {:ok, username}
+      _ ->
+        Logger.warn("Invalid basic auth credentials in request")
+        {:error, :invalid_credentials}
+    end
+  end
+
+  defp config(key) do
+    Application.get_env(:sanbase, __MODULE__)
```

I think it's better to use fetch_env! as it will raise an error if the env variable is not set and not return a default value